[DispatchCreation] Modify the generated fused op to not use concats.

MaheshRavishankar opened the pull request:
This is an almost complete rewrite of the  pass to fuse contractions horizontally which instead of concatenating operands to map to a GEMM, followed by slices to extract the individual matmul results; the pass now just creates a new operation with the operands being the common LHS, the RHS of each of the gemms, and the output of each of the gemms. The generated op yields the result of each constituent matmul.
This also allows for the RHS/output indexing maps of the gemms to be mismatched, since only the LHS operand and indexing maps need to match. The change also permutes the iteration space of the gemms to ensure that the same indexing maps are used for the LHS across all the fused matmuls.

The rest of the compiler stack has already been fixed up to handle such operations.

MaheshRavishankar committed:
[DispatchCreation] Modify the generated fused op to not use concats.

The change also allows doing horizontal fusion in cases where the LHS
operand is the same, but the RHS/Outputs might be transposed.

Signed-off-by: MaheshRavishankar <mravisha@amd.com>

IanWood1 commented on compiler/src/iree/compiler/DispatchCreation/FuseHorizontalContractions.cpp:
```diff
@@ -283,336 +220,247 @@ static std::optional<HorizontalFusionGroup> getHorizontalFusionGroupMembers(
+  for (auto linalgOp : linalgOps) {
+    rewriter.replaceOp(linalgOp, fusedOp->getResults().slice(
+                                     resultsIndex, linalgOp->getNumResults()));
+    resultsIndex += linalgOp->getNumResults();
```
The use of `linalgOp` after `replaceOp` should be the cause of the sanitizer failure.

IanWood1 commented on compiler/src/iree/compiler/DispatchCreation/FuseHorizontalContractions.cpp:
```diff
@@ -283,336 +220,247 @@ static std::optional<HorizontalFusionGroup> getHorizontalFusionGroupMembers(
+/// fusion to make sure all the LHS operands of the horizontally fused
+/// ops have the same indexing map.
+LogicalResult
+permuteIndexingMapsToMatchSeedLhs(RewriterBase &rewriter,
```
```suggestion
static LogicalResult
permuteIndexingMapsToMatchSeedLhs(RewriterBase &rewriter,
```

IanWood1 commented on compiler/src/iree/compiler/DispatchCreation/FuseHorizontalContractions.cpp:
```diff
@@ -283,336 +220,247 @@ static std::optional<HorizontalFusionGroup> getHorizontalFusionGroupMembers(
+template <typename V, typename R>
+static void appendRange(V &vector, R &&range) {
+  vector.append(range.begin(), range.end());
+}
```
`llvm::append_range` should do this right?

MaheshRavishankar replied:
TIL

MaheshRavishankar committed:
Avoid horizontal fusion when n dimensions mismatch.

Signed-off-by: MaheshRavishankar <mravisha@amd.com>

MaheshRavishankar committed:
Address comments (2).

Signed-off-by: MaheshRavishankar <mravisha@amd.com>

MaheshRavishankar committed:
Drop horizontal fusion flag from SD3 since the backend doesnt support this yet.

Previous implementation of horizontal fusion missed opportunities for
horizontal fusion in SD3, but now they do get picked up, but the
backend doesnt work on these. Dropping the flag is a no-op for the
test since there was no horizontal fusion to start with.

Signed-off-by: MaheshRavishankar <mravisha@amd.com>

qedawkins commented on compiler/src/iree/compiler/DispatchCreation/test/fuse_horizontal_contractions.mlir:
```diff
@@ -294,11 +247,121 @@ util.func public @dont_fuse_when_same_trunc_op(%arg0: tensor<2x4096x640xi8>, %ar
+//  CHECK-SAME:       ins(%[[ARG0]], %[[ARG1]], %[[ARG4]] :
+//       CHECK:       outs(%[[FILL]], %[[FILL]] :
+//       CHECK:   %[[TRUNCF:.+]] = linalg.generic
+//  CHECK-SAME:       ins(%[[FUSED_OP]]#1, %[[FUSED_OP]]#0 :
```
I don't think we want to horizontally fuse the truncates like this. Instead we should be able to keep them separate and just put them in the same dispatch. That will keep tile + fuse simpler at the workgroup level.

MaheshRavishankar replied:
I have an issue for fixing it at the workgroup level. This functionality is needed for llama ffn layer horizontal fusion.  https://github.com/llvm/llvm-project/pull/125915 is meant to fix it, but I havent reviewed it yet.

MaheshRavishankar replied:
Also, just to clarify. The trunc operator is not fused. It is an elementwise op that using results of both the matmuls.

qedawkins replied:
oooh that code looks spooky. I'm not sure how that's supposed to work without changing how we decide whether the fusion is legal.

> The trunc operator is not fused. It is an elementwise op that using results of both the matmuls

Exactly, so what is the benefit of putting them all in a single op? I would expect multiple different elementwise ops to work the same way.

MaheshRavishankar replied:
cause they will be created and consumed in the same dispatch. I agree I need to actually work with it to see how this works.

hanhanW commented on compiler/src/iree/compiler/DispatchCreation/FuseHorizontalContractions.cpp:
```diff
@@ -47,62 +48,69 @@ struct FuseHorizontalContractionsPass final
+         linalg::isaContractionOpInterface(bLinalgOp) &&
+         "expected lhs and rhs to be contraction ops");
+
+  // CHeck that the LHS operand is the same. Can be relaxed in the future.
```
nit: `Check`

hanhanW commented on compiler/src/iree/compiler/DispatchCreation/FuseHorizontalContractions.cpp:
```diff
@@ -47,62 +48,69 @@ struct FuseHorizontalContractionsPass final
-    std::optional<linalg::LinalgOp> seedContractionOp = std::nullopt) {
-  if (!linalg::isaContractionOpInterface(linalgOp)) {
+  // Check that the output rank and element type are the same. We dont check the
+  // type cause we allow RHS to be transposes.
```
I don't follow this. How does transposed RHS impact the output type? Is it a copy-paste typo?

MaheshRavishankar replied:
yeah, copy-pasta.

hanhanW commented on compiler/src/iree/compiler/DispatchCreation/FuseHorizontalContractions.cpp:
```diff
@@ -47,62 +48,69 @@ struct FuseHorizontalContractionsPass final
-    }
-    return success(equivalentValues.lookup(lhsValue) == rhsValue ||
-                   equivalentValues.lookup(rhsValue) == lhsValue);
+static bool checkOperationEquivalence(Operation *aOp, Operation *bOp) {
```
The function is tied to contraction in the implementation details, and it is assuming the operations all implement linalg::ContractionOpInteface. Perhaps we can rename it as `checkContractionEquivalence`?

hanhanW commented on compiler/src/iree/compiler/DispatchCreation/FuseHorizontalContractions.cpp:
```diff
@@ -47,62 +48,69 @@ struct FuseHorizontalContractionsPass final
-    return checkOperationEquivalence(linalgOp, seedContractionOp.value());
+    return checkOperationEquivalence(seedContractionOp.value(), linalgOp);
   }
   return true;
```
How about we inline the `checkOperationEquivalence` logic to here. This function is fairly small to me, so I think we do not need an additional function call. E.g., it can be

```cpp
if (!seedContractionOp) {
  return true;
}
// the logic of checkOperationEquivalence
```

MaheshRavishankar replied:
I want to leave it this way...  The `checkContractionEquivalence` already assumes the two ops are contractions.

hanhanW commented on compiler/src/iree/compiler/DispatchCreation/FuseHorizontalContractions.cpp:
```diff
@@ -47,62 +48,69 @@ struct FuseHorizontalContractionsPass final
-  if (!matchPattern(fillOp.getDpsInputOperand(0)->get(), m_AnyZeroFloat()) &&
-      !matchPattern(fillOp.getDpsInputOperand(0)->get(), m_Zero())) {
+
+  // Check region equivalence.
```
out of curiosity, is it a hard limit, or it can be relaxed in the future as well?

MaheshRavishankar replied:
I think it could be relaxed... but starting with this for now.

hanhanW commented on compiler/src/iree/compiler/DispatchCreation/FuseHorizontalContractions.cpp:
```diff
@@ -283,336 +220,247 @@ static std::optional<HorizontalFusionGroup> getHorizontalFusionGroupMembers(
+  SmallVector<AffineMap> fusedInsIndexingMaps;
+  SmallVector<AffineMap> fusedOutsIndexingMaps;
+
+  linalg::LinalgOp seedOp = cast<linalg::LinalgOp>(linalgOps.front());
```
style nit: use `auto` because `cast` already spells the type.

https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable

hanhanW commented on compiler/src/iree/compiler/DispatchCreation/FuseHorizontalContractions.cpp:
```diff
@@ -283,336 +220,247 @@ static std::optional<HorizontalFusionGroup> getHorizontalFusionGroupMembers(
+
+  linalg::LinalgOp seedOp = cast<linalg::LinalgOp>(linalgOps.front());
+  SmallVector<utils::IteratorType> fusedIteratorTypes =
+      cast<linalg::LinalgOp>(linalgOps.front()).getIteratorTypesArray();
```
I think you can use the `seedOp`.

hanhanW commented on compiler/src/iree/compiler/DispatchCreation/FuseHorizontalContractions.cpp:
```diff
@@ -283,336 +220,247 @@ static std::optional<HorizontalFusionGroup> getHorizontalFusionGroupMembers(
+static void fuseGroup(RewriterBase &rewriter,
+                      MutableArrayRef<Operation *> fusionGroup,
+                      DominanceInfo &dominanceInfo) {
+  if (!llvm::all_of(fusionGroup, [](Operation *op) {
+        return isa_and_nonnull<linalg::LinalgOp>(op);
+      })) {
+    return;
   }
```
I recently learned that we can write it this way, which looks better to me.

```
if (!llvm::all_of(fusionGroup, llvm::IsaPred<linalg::LinalgOp>)) {
  return;
}
```

MaheshRavishankar replied:
I think `llvm::IsaPred`  doesnt check for the operation not being null (i.e. null value returns true). I want `isa_and_nonnull` (though the op cant be null but just being careful).

hanhanW replied:
If it can not be a null, it triggers the assertion in `isa`. So I think it is still cleaner with `IsaPred`. Instead of handling the issue silently, the debug build captures the failure for you.

MaheshRavishankar replied:
I dont want it to assert though. I want it to return false if it is null.

hanhanW commented on compiler/src/iree/compiler/DispatchCreation/FuseHorizontalContractions.cpp:
```diff
@@ -283,336 +220,247 @@ static std::optional<HorizontalFusionGroup> getHorizontalFusionGroupMembers(
+    return;
   }
-  return success();
+  linalg::LinalgOp baseContractOp = cast<linalg::LinalgOp>(fusionGroup.front());
```
ditto: use `auto`

hanhanW reviewed (commented):
We are (already) missing the documentation of the pass. The PR description looks good to me. Can you add such documentation to `Passes.td`? I.e., add a `description` section to the pass definition.

https://github.com/iree-org/iree/blob/6ebfcaa9186b3d8fc4d5191d653978ee2176f6d9/compiler/src/iree/compiler/DispatchCreation/Passes.td#L68-L84

MaheshRavishankar committed:
Fix partial horizontal fusion dependence violation.

Signed-off-by: MaheshRavishankar <mravisha@amd.com>

MaheshRavishankar committed:
Address comments (3)

Signed-off-by: MaheshRavishankar <mravisha@amd.com>

hanhanW commented on compiler/src/iree/compiler/DispatchCreation/FuseHorizontalContractions.cpp:
```diff
@@ -283,336 +252,243 @@ static std::optional<HorizontalFusionGroup> getHorizontalFusionGroupMembers(
+      getResultDimsRange(seedLhsIndexingMap.getResults());
+  auto lhsResultDimsRange = getResultDimsRange(lhsIndexingMap.getResults());
+
+  // Start with a identity permutations. For now try to only swap dimensions
```
nit:

```suggestion
  // Start with an identity permutation. For now try to only swap dimensions
```

hanhanW commented on compiler/src/iree/compiler/DispatchCreation/FuseHorizontalContractions.cpp:
```diff
@@ -283,336 +252,243 @@ static std::optional<HorizontalFusionGroup> getHorizontalFusionGroupMembers(
+    if (interchangeVector[lhsDimPos] == lhsDimPos &&
+        interchangeVector[seedDimPos] == seedDimPos) {
+      interchangeVector[lhsDimPos] = seedDimPos;
+      interchangeVector[seedDimPos] = lhsDimPos;
```
optional nit: I'd write `std::use` which maps to the comment directly.

MaheshRavishankar replied:
Whats std::use?

hanhanW replied:
Sorry, I meant `std::swap`..

MaheshRavishankar replied:
good point.

hanhanW commented on compiler/src/iree/compiler/DispatchCreation/FuseHorizontalContractions.cpp:
```diff
@@ -283,336 +252,243 @@ static std::optional<HorizontalFusionGroup> getHorizontalFusionGroupMembers(
+  SmallVector<utils::IteratorType> permutedIteratorTypes =
+      llvm::to_vector(iteratorTypes);
+  if (permutedIteratorTypes != iteratorTypes) {
+    return failure();
```
What is this check for? I think you want to check whether `applyPermutation(origIteratorTypes, interchangeVector)` is as the same as `iteratorTypes` or not. However, the `origIteratorTypes` is not passed in?

MaheshRavishankar replied:
Haha! Good catch! I have already checked earlier that the iterator types match to start with. All I need to check is that after permutation, the iterator types remains the same. I fixed it now.

hanhanW commented on compiler/src/iree/compiler/DispatchCreation/FuseHorizontalContractions.cpp:
```diff
@@ -283,336 +252,243 @@ static std::optional<HorizontalFusionGroup> getHorizontalFusionGroupMembers(
+
+  auto seedOp = cast<linalg::LinalgOp>(linalgOps.front());
+  SmallVector<utils::IteratorType> fusedIteratorTypes =
+      cast<linalg::LinalgOp>(seedOp).getIteratorTypesArray();
```
I meant that you can write `seedOp.getIteratorTypesArray()` without casting.

hanhanW commented on compiler/src/iree/compiler/DispatchCreation/FuseHorizontalContractions.cpp:
```diff
@@ -283,336 +252,243 @@ static std::optional<HorizontalFusionGroup> getHorizontalFusionGroupMembers(
+/// that the LHS of the contraction operations fused horizontally is the same
+/// and have the same indexing map for all the operations. The RHS/outputs of
+/// the operations can be different.
+static std::optional<linalg::GenericOp>
+fuseContractionsHorizontally(RewriterBase &rewriter, Location loc,
+                             MutableArrayRef<Operation *> linalgOps) {
```
The returned value is not used at all. If the intention is dumping it for debugging, I think you can log it with `LLVM_DEBUG` within the function. I.e., we can make it return void.

The other question is that why is `linalgOps` mutable? Is it because that they will be replaced with the new results at the end?

MaheshRavishankar replied:
I think it is good practice to return the newly created op from these kind of transformations. The caller can decide to drop it or not. So its kind of a "norm" in my mind. 
The reason for `MutableArrayRef<Operation *>` was that calling some methods would give me a compile-time error since `ArrayRef<Operation *>` somehow makes all methods go to `const ` method of the underlying object.

hanhanW replied:
Okay, can you document the return value in this case? Returning `std::nullopt` is not a failure, and I think we should document it.

hanhanW commented on compiler/src/iree/compiler/DispatchCreation/FuseHorizontalContractions.cpp:
```diff
@@ -283,336 +252,243 @@ static std::optional<HorizontalFusionGroup> getHorizontalFusionGroupMembers(
+
+  Block *fusedBody = fusedOp.getBlock();
+  auto rhsIndex = 0;
+  auto outsIndex = fusedOp.getNumDpsInputs();
```
style nit: spell out the type.

> Use type deduction only if it makes the code clearer to readers who aren't familiar with the project, or if it makes the code safer. Do not use it merely to avoid the inconvenience of writing an explicit type.

https://google.github.io/styleguide/cppguide.html#Type_deduction

>  Use auto if and only if it makes the code more readable or easier to maintain.

https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable

hanhanW commented on compiler/src/iree/compiler/DispatchCreation/FuseHorizontalContractions.cpp:
```diff
@@ -283,336 +252,243 @@ static std::optional<HorizontalFusionGroup> getHorizontalFusionGroupMembers(
+
+  auto resultsIndex = 0;
+  for (auto linalgOp : linalgOps) {
+    auto numResults = linalgOp->getNumResults();
```
ditto, spell out the type for `resultsIndex` and `numResults`.

hanhanW commented on compiler/src/iree/compiler/DispatchCreation/FuseHorizontalContractions.cpp:
```diff
@@ -283,336 +252,243 @@ static std::optional<HorizontalFusionGroup> getHorizontalFusionGroupMembers(
-                               newTruncOp.getRegion().begin());
-
-    concatResultOp = cast<linalg::LinalgOp>(newTruncOp.getOperation());
+    // Append the RHS operands;
```
nit: typo

```suggestion
    // Append the RHS operands.
```

hanhanW commented on compiler/src/iree/compiler/DispatchCreation/FuseHorizontalContractions.cpp:
```diff
@@ -171,52 +179,25 @@ getTruncateOp(Operation *op,
-/// accumulate in higher precision, followed by a `linalg.generic` that performs
-/// the `truncf` on the result.
-static std::optional<HorizontalFusionGroup> getHorizontalFusionGroupMembers(
+/// Note: The actual operation generated does not concat the RHS
```
nit: add a period at the end to finish the sentence

```suggestion
/// Note: The actual operation generated does not concat the RHS.
```

hanhanW reviewed (commented):
Looks good, just few nits + a question about a check.

qedawkins commented on compiler/src/iree/compiler/DispatchCreation/FuseHorizontalContractions.cpp:
```diff
@@ -47,62 +58,90 @@ struct FuseHorizontalContractionsPass final
+  if (failed(aContactionDims) || failed(bContactionDims)) {
+    return false;
+  }
+  if (aContactionDims.value() != bContactionDims.value()) {
```
nit: typo
```suggestion
  if (aContractionDims.value() != bContractionDims.value()) {
```

qedawkins commented on compiler/src/iree/compiler/DispatchCreation/FuseHorizontalContractions.cpp:
```diff
@@ -47,62 +58,90 @@ struct FuseHorizontalContractionsPass final
+/// Check that an operation is a `contraction`
+static bool isEquivalentContractionOp(
+    linalg::LinalgOp linalgOp,
+    std::optional<linalg::LinalgOp> seedContractionOp = std::nullopt) {
```
We can just drop the optionality and call `linalg::isaContractionOpInterface` directly in the one place that uses this.

MaheshRavishankar replied:
Hanhan asked for that too. I just deleted this method and inlined its implementation.

qedawkins commented on compiler/src/iree/compiler/DispatchCreation/FuseHorizontalContractions.cpp:
```diff
@@ -283,336 +252,243 @@ static std::optional<HorizontalFusionGroup> getHorizontalFusionGroupMembers(
+/// results, corresponding to the results of the fused operations. It is assumed
+/// that the LHS of the contraction operations fused horizontally is the same
+/// and have the same indexing map for all the operations. The RHS/outputs of
+/// the operations can be different.
```
```suggestion
/// the operations can be different, but must share the same shape (and thus iteration space).
```

qedawkins reviewed (approved):
Cool, LGTM % nits!

MaheshRavishankar committed:
Address comments (4)

Signed-off-by: MaheshRavishankar <mravisha@amd.com>

MaheshRavishankar reviewed (commented):
Thanks for the reviews!

hanhanW commented on compiler/src/iree/compiler/DispatchCreation/FuseHorizontalContractions.cpp:
```diff
@@ -283,336 +240,243 @@ static std::optional<HorizontalFusionGroup> getHorizontalFusionGroupMembers(
+    return;
+  }
+
+  fuseContractionsHorizontally(rewriter, loc, fusionGroup);
```
[optional]: I'm being careful here. It is better to have void casting or `[[maybe_unused]]`. Otherwise, some build systems might complain that the return value is ignored as a warning; it becomes a build failure if they enable `-Werror`. This might not be an issue in IREE CI, but it could happen in downstream projects if they do not set the build system properly.

hanhanW commented on compiler/src/iree/compiler/DispatchCreation/FuseHorizontalContractions.cpp:
```diff
@@ -283,336 +240,243 @@ static std::optional<HorizontalFusionGroup> getHorizontalFusionGroupMembers(
-                               newTruncOp.getRegion().begin());
-
-    concatResultOp = cast<linalg::LinalgOp>(newTruncOp.getOperation());
+    // Append the RHS operands.
+    SmallVector<OpOperand *> ins = linalgOp.getDpsInputOperands();
+    llvm::append_range(
+        fusedIns,
+        llvm::map_range(ArrayRef<OpOperand *>(ins).drop_front(),
+                        [](OpOperand *operand) { return operand->get(); }));
+
+    // Append the Outs operands.
+    llvm::append_range(fusedOuts, llvm::map_range(linalgOp.getDpsInitsMutable(),
+                                                  [](OpOperand &operand) {
+                                                    return operand.get();
+                                                  }));
+
+    // Append the result types.
+    fusedResultTypes.append(linalgOp->result_type_begin(),
+                            linalgOp->result_type_end());
+
+    // Append the rhs indexing maps.
+    llvm::append_range(fusedInsIndexingMaps,
+                       ArrayRef<AffineMap>(opIndexingMaps)
+                           .slice(1, linalgOp.getNumDpsInputs() - 1));
+
+    // Append the outs indexing maps.
+    llvm::append_range(fusedOutsIndexingMaps,
+                       ArrayRef<AffineMap>(opIndexingMaps)
+                           .drop_front(linalgOp.getNumDpsInputs()));
```
I'm okay with the change. I just want to raise that they can be simpler (e.g., a single push_back) if the assumption is that they are all ContractionOpInterface ops. By [definition](https://github.com/llvm/llvm-project/blob/2b41277a09820bc47dc533ad37b5213edc2e8d52/mlir/lib/Dialect/Linalg/IR/LinalgInterfaces.cpp#L483-L484), they all have two input operands and one init operand. I marked it optional because the definition could change in the future.

MaheshRavishankar replied:
I actually started with something very general, and then had to work my way back to something more constrained. So yo uare right, but i'd rather leave it this way for now.

hanhanW reviewed (approved):
LG, just some optional nits.
